ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/LedgerDao.scala

```diff
@@ -193,6 +196,13 @@ trait LedgerWriteDao extends AutoCloseable {
     */
   def initializeLedger(ledgerId: LedgerId, ledgerEnd: LedgerOffset): Future[Unit]
 
+  /**
+    * Initializes the participant. Must be called only once.
```

Comment:
What happens if it's called multiple times?

Reply:
Undefined behavior, but as of right now it will insert a second row into `parameters`. We should probably make the `ledger_id` column the primary key of the table, so that it fails.

Reply:
I would add some kind of check. The parameters table should always contain 0 or 1 rows.
- `initializeLedger` uses `INSERT INTO parameters`. It behaves unexpectedly if there already is one row (i.e., initializeLedger has already been called).
- `initializeParticipant` uses `UPDATE parameters`. It behaves unexpectedly if there is no row (i.e., initializeLedger has not yet been called).

I can see someone accidentally swapping the calls of `initializeLedger` and `initializeParticipant` in the future.

We could manually check the number of rows in `initializeLedger` and `initializeParticipant`. `initializeParticipant` could use `executeUpdate()`, which returns the number of rows affected.

Another idea is to add a column `id INT PRIMARY KEY NOT NULL DEFAULT(1) CHECK (id = 1)` on the parameters table, which enforces the fact that there is never more than one row.

Reply:
I agree it's not optimal. But since we only now introduce the participant id, we need to differentiate between starting from an empty database (no entry in parameters, both ledgerId and participantId need to be set) or an existing ledger (we only need set the participant id).